test/test_autograd.py

```diff
@@ -6207,6 +6205,70 @@ def get_refs(with_backward):
         gc.collect()
         self.assertTrue(ref.expired())
 
+    def test_create_graph_and_full_backward_hook_cycle(self):
+        # If BackwardHook saves grad_output, it can create a cycle when we perform backward
+        # with create_graph=True
+        #
+        #   grad_output -> grad_output.grad_fn -> graph -> hook -> grad_output
+        #
+        class TestCls():
+            # Dummy class for the purpose of creating a weakref
+            pass
+
+        def get_ref(input_requires_grad, nb_hooks, remove_hook_during_backward):
+            t = torch.randn(10, requires_grad=input_requires_grad)
+            a = torch.tensor(1., requires_grad=True)
+
+            def f(*args):
+                # Removing the hooks during backward doesn't matter -
+                # All the hooks will still be executed, and counter will still go to zero
+                if remove_hook_during_backward:
+                    handles[0].remove()
+                return None
+
+            class Test(nn.Module):
+                def forward(self, x):
+                    tmp = a ** 2
+                    # register a hook to try to remove the full backward hook during backward
+                    tmp.grad_fn.register_hook(f)
```

Comment:
This is a post hook on tmp.grad_fn ? Is that chosen on purpose?
it is a bit unclear to me what you're trying to do here?

Reply:
`f` is supposed to run after we set `self.grad_output`, but before we call the full backward hook (registered to the dummy nodes on the args). If we are able to remove a hook at that time, the counter would not reach zero.

I don't know if we need this any more if we go with the plan of moving the loop, but yeah maybe registering it to the output of `a ** 2 * x ** 2` would've been more accurate.